filemanager/s3manager_v2.go

```diff
@@ -0,0 +1,366 @@
+	}
+	uploader := s3manager.NewUploader(client)
+
+	ctx, cancel := context.WithTimeout(ctx, m.getTimeout())
+	defer cancel()
+
+	output, err := uploader.Upload(ctx, uploadInput)
+	if err == nil {
+		return UploadedFile{Location: output.Location, ObjectName: fileName}, nil
+	}
+	var regionError *aws.MissingRegionError
+	if errors.As(err, &regionError) {
+		err = fmt.Errorf(`missing region for bucket %q: %w`, m.config.Bucket, regionError)
+	}
+
+	return UploadedFile{}, err
+}
+
+// UploadReader uploads data from an io.Reader to S3 with the given object name.
+func (m *S3ManagerV2) UploadReader(ctx context.Context, objName string, rdr io.Reader) (UploadedFile, error) {
+	if objName == "" {
+		return UploadedFile{}, errors.New("object name cannot be empty")
+	}
+	fileName := path.Join(m.config.Prefix, objName)
+
+	uploadInput := &s3.PutObjectInput{
+		ACL:    types.ObjectCannedACLBucketOwnerFullControl,
+		Bucket: aws.String(m.config.Bucket),
+		Key:    aws.String(fileName),
+		Body:   rdr,
+	}
+
+	if m.config.EnableSSE {
+		uploadInput.ServerSideEncryption = types.ServerSideEncryptionAes256
+	}
+
+	client, err := m.getClient(ctx)
+	if err != nil {
+		return UploadedFile{}, fmt.Errorf("s3 client: %w", err)
+	}
+	uploader := s3manager.NewUploader(client)
+
+	ctx, cancel := context.WithTimeout(ctx, m.getTimeout())
+	defer cancel()
+
+	output, err := uploader.Upload(ctx, uploadInput)
+	if err == nil {
+		return UploadedFile{Location: output.Location, ObjectName: fileName}, nil
+	}
+	var regionError *aws.MissingRegionError
+	if errors.As(err, &regionError) {
+		err = fmt.Errorf(`missing region for bucket %q: %w`, m.config.Bucket, regionError)
+	}
+	return UploadedFile{}, err
+}
+
+// Delete removes the specified keys from S3.
+func (m *S3ManagerV2) Delete(ctx context.Context, keys []string) error {
+	client, err := m.getClient(ctx)
+	if err != nil {
+		return fmt.Errorf("s3 client: %w", err)
+	}
+
+	if len(keys) == 0 {
+		return nil
+	}
+
+	var objects []types.ObjectIdentifier
+	for _, key := range keys {
+		objects = append(objects, types.ObjectIdentifier{Key: aws.String(key)})
+	}
+
+	batchSize := 1000 // max accepted by DeleteObjects API
+	chunks := lo.Chunk(objects, batchSize)
+	for _, chunk := range chunks {
+		deleteCtx, cancel := context.WithTimeout(ctx, m.getTimeout())
+		_, err := client.DeleteObjects(deleteCtx, &s3.DeleteObjectsInput{
+			Bucket: aws.String(m.config.Bucket),
+			Delete: &types.Delete{
+				Objects: chunk,
+			},
+		})
+		cancel()
+		if err != nil {
+			var apiErr smithy.APIError
+			if errors.As(err, &apiErr) {
+				m.logger.Errorn("Error while deleting S3 objects", logger.NewErrorField(err), logger.NewStringField("error_code", apiErr.ErrorCode()))
+			} else {
+				m.logger.Errorn("Error while deleting S3 objects", logger.NewErrorField(err))
+			}
+			return fmt.Errorf("failed to delete S3 objects: %w", err)
+		}
+	}
+	return nil
+}
+
+// Prefix returns the configured S3 prefix.
+func (m *S3ManagerV2) Prefix() string {
```

Comment:
do we need the `Bucket` method as well?

Reply:
We don't need a Bucket either in the v1 manager so removing that

Reply:
what about this?
https://github.com/rudderlabs/rudder-sources/blob/2028d1687411dd41cf104371e477bc399f042621/integrations/replay/internal/errorindex/filemanager_impl.go#L95